api/server/sdk/alerts.go

```diff
@@ -0,0 +1,157 @@
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/pkg/proto/time"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// alertsServer implements api.OpenStorageAlertsServer
+type alertsServer struct {
+	// reader holds pointer to alerts reader
+	reader alerts.Reader
+}
+
+func getOpts(opts []*api.SdkAlertsOption) []alerts.Option {
+	var options []alerts.Option
+
+	for _, opt := range opts {
+		switch opt.GetOpt().(type) {
+		case *api.SdkAlertsOption_MinSeverityType:
+			options = append(options,
+				alerts.NewMinSeverityOption(opt.GetMinSeverityType()))
+		case *api.SdkAlertsOption_IsCleared:
+			options = append(options,
+				alerts.NewFlagCheckOption(opt.GetIsCleared()))
+		case *api.SdkAlertsOption_TimeSpan:
+			options = append(options,
+				alerts.NewTimeSpanOption(
+					prototime.TimestampToTime(opt.GetTimeSpan().GetStartTime()),
+					prototime.TimestampToTime(opt.GetTimeSpan().GetEndTime())))
+		case *api.SdkAlertsOption_CountSpan:
+			options = append(options,
+				alerts.NewCountSpanOption(
+					opt.GetCountSpan().GetMinCount(),
+					opt.GetCountSpan().GetMaxCount()))
+		}
+	}
+
+	return options
+}
+
+func getFilters(queries []*api.SdkAlertsQuery) []alerts.Filter {
+	var filters []alerts.Filter
+
+	// range over all queries
+	for _, x := range queries {
+		switch x.GetQuery().(type) {
+		case *api.SdkAlertsQuery_ResourceTypeQuery:
+			q := x.GetResourceTypeQuery()
+			filters = append(filters,
+				alerts.NewResourceTypeFilter(
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		case *api.SdkAlertsQuery_AlertTypeQuery:
+			q := x.GetAlertTypeQuery()
+			filters = append(filters,
+				alerts.NewAlertTypeFilter(
+					q.AlertType,
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		case *api.SdkAlertsQuery_ResourceIdQuery:
+			q := x.GetResourceIdQuery()
+			filters = append(filters,
+				alerts.NewResourceIDFilter(
+					q.ResourceId,
+					q.AlertType,
+					q.ResourceType,
+					getOpts(x.GetOpts())...))
+		}
+	}
+
+	return filters
+}
+
+// Enumerate implements api.OpenStorageAlertsServer for alertsServer.
+// Input context should ideally have a deadline, in which case, a
+// graceful exit is ensured within that deadline.
+func (g *alertsServer) Enumerate(ctx context.Context,
+	request *api.SdkAlertsEnumerateRequest) (*api.SdkAlertsEnumerateResponse, error) {
+
+	// if input has deadline, ensure graceful exit within that deadline.
+	deadline, ok := ctx.Deadline()
+	var cancel context.CancelFunc
+	if ok {
+		// create a new context that will get done on deadline
+		ctx, cancel = context.WithTimeout(ctx, deadline.Sub(time.Now()))
+		defer cancel()
+	}
+
+	group, _ := errgroup.WithContext(ctx)
+	errChan := make(chan error)
+
+	resp := new(api.SdkAlertsEnumerateResponse)
+	var mu sync.Mutex
+
+	queries := request.GetQueries()
+	if queries == nil {
+		return nil, status.Error(codes.InvalidArgument, "input is nil")
```

Comment:
Need to be more clear with the error output. It is seen by developers and needs to be clear. Something like: `Must provide a query`.

Reply:
done